### PR TITLE
mgr: use annotations for argdesc

### PR DESCRIPTION
**Pull request by tchaikov**
this changeset is a superset of #38718. 

#38718 uses docstring for command description (the help message). and this change goes further. it uses annotations for command argument description as well.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -703,24 +704,23 @@ def _set_ssh_config(self, inbuf: Optional[str] = None) -> Tuple[int, str, str]:
...
         self.ssh_config_tmp = None
         self.log.info('Cleared ssh_config')
         self._reconfig_ssh()
         return 0, "", ""
 
-    @orchestrator._cli_read_command(
-        prefix='cephadm get-ssh-config',
-        desc='Returns the ssh config as used by cephadm'
-    )
+    @orchestrator._cli_read_command('cephadm get-ssh-config')
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:719]: we still have all API docs being broken on docs.ceph.com. I don't know if this change is going to break the automatically generated mon command API t(hat is broken right now anyway)

File: src/pybind/mgr/orchestrator/_interface.py [hunk 2]
```diff
@@ -102,15 +102,16 @@ def wrapper(*args, **kwargs):
     # misuse partial to copy `wrapper`
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/_interface.py:102]:
```suggestion
    # misuse lambda to copy `wrapper` function
```

File: src/pybind/mgr/orchestrator/module.py [hunk 3]
```diff
@@ -29,7 +30,41 @@ def nice_delta(now, t, suffix=''):
...
+    mon = 'mon'
+    mgr = 'mgr'
+    rbd_mirror = 'rbd-mirror'
+    crash = 'crash'
+    alertmanager = 'alertmanager'
+    grafana = 'grafana'
+    node_exporter = 'node-exporter'
+    prometheus = 'prometheus'
+    cephadm_exporter = 'cephadm-exporter'
+
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:50]: https://github.com/ceph/ceph/blob/a51f8f2612d5815a51d4ab242a9a90c153155686/src/python-common/ceph/deployment/service_spec.py#L382-L384

**Review by sebastian-philipp — COMMENTED** [review]
looks great!

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (tchaikov) [reply src/pybind/mgr/cephadm/module.py:719]: @sebastian-philipp what do you mean by "all"? i think the only missing pieces of the API docs are 

- document generated using javadoc
- mon command doc generated using `gen_mon_command_api.py`. this part is missing from docs.ceph.com since its day 1. see https://github.com/ceph/ceph/pull/33886#issuecomment-613430619 .

tl;dr. yes. it's going to break the mon + mgr commands doc build.

long story: since `src/script/gen_static_command_descriptions.py` imports mgr modules and parses the `cmd` attribute of the elements in `COMMANDS`, and `cmd` is composed of `prefix` and `args` attributes of a command. the latter is rebuilt when a command is registered using the `CLICommand` decorator and its variants. so presumably, it won't be difficult to adapt this script to use `prefix` and `args` instead of `cmd`.

if you are fine with the direction of this change in general. i will do the adaption work. but i don't have plan to fix the broken mon command doc at this moment. i did some investigations though. a viable approach is to create a homebrew sphinx extension for generating the sphinx docs (not the rst files) from the code, like what autodoc does.

File: src/pybind/mgr/orchestrator/module.py (hunk 3 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/module.py:50]: i'd fix `python-common/ceph/deployment/service_spec.py` then, so it's typed. and can be fit into the annotation system. or do you suggest we should use `Literal` ?

File: src/ceph.in [hunk 4]
```diff
@@ -257,7 +259,9 @@ GLOBAL_ARGS = {
 }
 
 
-def parse_cmdargs(args=None, target=''):
+def parse_cmdargs(args=None, target='') -> Tuple[argparse.ArgumentParser,
+                                                 argparse.Namespace,
+                                                 List[str]]:
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/ceph.in:264]: i think the first step is to add more (enough) annotations to ceph.in and some of the homebrew python modules it is using, so i can fix some of the obvious issues. and then run mypy against it.

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/_interface.py:102]: sure, will create a separate commit to address this.

**Comment by tchaikov** [comment]
rebased against master.

File: src/ceph.in (hunk 4 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/ceph.in]: FWIW, this part has been merged as #38713.

[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/ceph.in]: If you're interested, I once had a bigger PR #26769 with lots of type hints. Many of them are still not yet part of master.

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py:719]: for gen_mon_command_api, we could also simply commit the generated .rst file into the tree. We just need a make check test to verify the file is up to date. wdyt?

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/_interface.py:102]: please don't worry too much. I simply missed adapting this code when trying out multiple different implementations and forgot to update this comment after being done. 

I simply though it would make sense to resolve this, as you're touching the code anyway.

File: src/pybind/mgr/orchestrator/module.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/module.py:50]: Your introduction of enums is awesome. I really like them. In fact my plan would be to move them into the orchestrator interface an make use of them in mgr/cephadm later.

File: src/ceph.in (hunk 4 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/ceph.in]: thanks for pointing me to it, i can use it as a reference along the way.

**Comment by sebastian-philipp** [comment]
do you want to split this PR into 

1. provide the infrastructure
2. change the modules

or do you plan to merge this in one go?

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (tchaikov) [reply src/pybind/mgr/cephadm/module.py:719]: #38807

**Comment by tchaikov** [comment]
@sebastian-philipp what do you mean by "splitting"? this change is split into multiple commits. the commits before "mgr/devicehealth: add arg annotations" does the ground work to prepare the infrastructure. the ones after "mgr/devicehealth: add arg annotations" add the annotations, later use the annotations for adding the argdesc and then remove the "arg" parameter. i can split the commits into smaller steps if you prefer.

**Comment by tchaikov** [comment]
> or do you plan to merge this in one go?

yes. I don't think it makes sense to split it into two or more PRs.

File: src/pybind/mgr/orchestrator/_interface.py (hunk 2 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/_interface.py]: add the commit for the stale comment.

**Comment by tchaikov** [comment]
@sebastian-philipp @epuertat this PR is ready for review. could you help review this changeset?

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test docs

**Comment by tchaikov** [comment]
jenkins test docs

File: src/pybind/mgr/orchestrator/module.py [hunk 5]
```diff
@@ -334,85 +364,81 @@ def _get_hosts(self, format='plain'):
...
-        "name=wide,type=CephBool,req=false",
-        'List devices on a host')
-    def _list_devices(self, hostname=None, format='plain', refresh=False, wide=False):
+    @_cli_read_command('orch device ls')
+    def _list_devices(self,
+                      hostname: Optional[str] = None,
+                      format: Optional[Format] = Format.plain,
+                      refresh: Optional[bool] = False,
+                      wide: Optional[bool] = False):
         # type: (Optional[List[str]], str, bool, bool) -> HandleCommandResult
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:421]: remove this line

File: src/pybind/mgr/orchestrator/module.py [hunk 6]
```diff
@@ -951,6 +950,7 @@ def _rgw_add(self,
                  ssl: bool = False,
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:950]:
```suggestion
                 ssl: Optional[bool] = False,
```

File: src/pybind/mgr/orchestrator/module.py [hunk 7]
```diff
@@ -1140,27 +1112,21 @@ def _apply_misc(self,
...
-        'name=dry_run,type=CephBool,req=false '
-        'name=unmanaged,type=CephBool,req=false '
-        'name=format,type=CephChoices,strings=plain|json|json-pretty|yaml,req=false',
-        'Update the number of MDS instances for the given fs_name')
+    @_cli_write_command('orch apply mds')
     def _apply_mds(self,
                    fs_name: str,
                    placement: Optional[str] = None,
                    dry_run: bool = False,
                    unmanaged: bool = False,
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:1126]:
```suggestion
                   dry_run: Optional[bool] = False,
                   unmanaged: Optional[bool] = False,
```

Btw (independent to this usage here), using `Optional[bool]` is typical an antipattern when using types, except if you **really** want to have `None` values in addition to `False` and `True`.  This is the reason I typically used standard non-optional bool annotations for optional bool parameters

File: src/pybind/mgr/orchestrator/module.py [hunk 8]
```diff
@@ -1232,31 +1188,23 @@ def _apply_rgw(self,
...
+    @_cli_write_command('orch apply nfs')
     def _apply_nfs(self,
                    svc_id: str,
                    pool: str,
                    namespace: Optional[str] = None,
                    placement: Optional[str] = None,
-                   format: str = 'plain',
+                   format: Optional[Format] = Format.plain,
                    dry_run: bool = False,
                    unmanaged: bool = False,
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:1205]:
```suggestion
                   dry_run: Optional[bool] = False,
                   unmanaged: Optional[bool] = False,
```

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]
only minor nits

File: src/pybind/mgr/dashboard/plugins/feature_toggles.py [hunk 9]
```diff
@@ -77,13 +78,10 @@ def get_options(self):
...
-            "dashboard feature",
-            "name=action,type=CephChoices,strings={} ".format(
-                "|".join(a.value for a in Actions))
-            + "name=features,type=CephChoices,strings={},req=false,n=N".format(
-                "|".join(f.value for f in Features)))
-        def cmd(mgr, action, features=None):
+        @CLICommand("dashboard feature")
+        def cmd(mgr,
+                action: Actions,
+                features: Optional[Sequence[Features]] = None):
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/dashboard/plugins/feature_toggles.py:84]: no, it's not. so i have
```py
            if orig_type in (abc.Sequence, Sequence, List, list):
                 assert len(type_args) == 1
                 attrs['n'] = 'N'
                 return CephArgtype.to_argdesc(type_args[0], attrs)
```
in src/pybind/ceph_argparse.py.

File: src/pybind/mgr/mgr_module.py [hunk 10]
```diff
@@ -286,57 +301,72 @@ def device_class_counts(self):
...
+
+    @staticmethod
+    def load_func_metadata(f):
+        desc = inspect.getdoc(f) or ''
+        full_argspec = inspect.getfullargspec(f)
+        arg_spec = full_argspec.annotations
+        KNOWN_ARGS = '_', 'self', 'mgr', 'inbuf'
+        for arg in full_argspec.args:
+            assert arg in KNOWN_ARGS or arg in arg_spec, \
+                f"'{arg}' is not annotated for {f}: {full_argspec}"
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/mgr_module.py]: thought it was more explicit by including `self` in the `KNOWN_ARGS`. we'd have to document `1` if we start with 1.

File: src/pybind/mgr/mgr_module.py [hunk 11]
```diff
@@ -286,57 +301,72 @@ def device_class_counts(self):
...
     def __call__(self, func):
+        self.store_func_metadata(func)
         self.func = func
-        if not self.desc:
-            self.desc = inspect.getdoc(func)
         self.COMMANDS[self.prefix] = self
         return self.func
 
-    def call(self, mgr, cmd_dict, inbuf):
+    def _get_arg_value(self, kwargs_switch, key, val):
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/mgr_module.py:335]: i can do this. but i don't expect user to use this internal function. i would want to prevent them from using it if i could.

File: src/pybind/mgr/mgr_module.py [hunk 12]
```diff
@@ -286,57 +301,72 @@ def device_class_counts(self):
...
+
+        if kwargs_switch:
+            k, v = val.split('=', 1)
+        else:
+            k, v = key, val
+        return kwargs_switch, k.replace('-', '_'), v
+
+    @staticmethod
+    def _is_optional(tp):
+        return get_origin(tp) is Union and type(None) in get_args(tp)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/mgr_module.py:354]: i don't really want to like have `not` unless necessary. it slows down the my buggy built-in parser a lot.

File: src/pybind/mgr/orchestrator/module.py (hunk 5 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/module.py:421]: done.

File: src/pybind/mgr/orchestrator/module.py (hunk 7 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/module.py:1126]: so the recommended way is plain `dry_run: bool = False`, right?  i can come up with a change which crosses check the `default` field and `annotations` field in a follow-up PR.

```
In [1]: def foo(dry: bool = False):
   ...:     pass
   ...:

In [2]: import inspect

In [3]: inspect.getfullargspec(foo)
Out[3]: FullArgSpec(args=['dry'], varargs=None, varkw=None, defaults=(False,), kwonlyargs=[], kwonlydefaults=None, annotations={'dry': <class 'bool'>})
```

File: src/pybind/mgr/orchestrator/module.py (hunk 8 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/module.py:1205]: done.

File: src/pybind/mgr/orchestrator/module.py (hunk 6 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/module.py:950]: done.

File: src/pybind/mgr/dashboard/plugins/debug.py [hunk 13]
```diff
@@ -55,14 +55,14 @@ def setup(self):
         self._refresh_health_checks()
 
     @no_type_check
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/dashboard/plugins/debug.py:57]: lemme try.

File: src/pybind/mgr/mgr_module.py [hunk 14]
```diff
@@ -286,57 +301,75 @@ def device_class_counts(self):
...
-                else:
-                    self.args_dict[v] = arg_d
+        self.arg_spec = {}    # type: Dict[str, Any]
+
+    KNOWN_ARGS = '_', 'self', 'mgr', 'inbuf'
+
+    @staticmethod
+    def load_func_metadata(f):
+        desc = inspect.getdoc(f) or ''
+        full_argspec = inspect.getfullargspec(f)
```
Comment (epuertat) [inline src/pybind/mgr/mgr_module.py:315]: What about inferring the command name? That way, `CLICommand` could probably be a decorator without args and infer everything from the function signature?

```suggestion
__module_prefix__ = 'dashboard'
...
    def load_func_metadata(f):
        if __module_prefix__:
            self.prefix = f'{__module_prefix__} {f.__name__.replace('_', '-')}'
        desc = inspect.getdoc(f) or ''
        full_argspec = inspect.getfullargspec(f)
```

**Review by epuertat — APPROVED** [review]
Looks awesome! Great job @tchaikov. Just a few comments over there!

**Review by epuertat — APPROVED** [review]
Looks awesome! Great job @tchaikov. Just a few comments over there!

File: src/pybind/ceph_argparse.py [hunk 15]
```diff
@@ -147,9 +161,52 @@ def __str__(self):
...
+    def __call__(self, v):
+        return v
+
     def complete(self, s):
         return []
 
+    @staticmethod
+    def to_argdesc(tp, attrs):
+        if tp is str:
+            return CephString().argdesc(attrs)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/ceph_argparse.py:173]: @epuertat very true. will make `argdesc()` a staticmethod. actually, i originally wanted to use `attrs` as a dict to initialize, for instance, `CephFloat`. currently we use the plain types like `int` and `string` to annotate the parameters. but there are cases where we want to use more strict validators for checking, for instance, the "range" of an `int`, the  "goodchars" of a `string`. in that case, we could use an instance of `CephString` to annotate the parameter, as it has builtin support to validate the "goodchar" regular expression. once we have the luxury to use Python 3.9 or we have a replacement of `typing.Annotated`, we can even use it to define our own parameter types and retire `CephString` and friends.

File: src/pybind/mgr/orchestrator/module.py (hunk 7 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator/module.py]: In my opinion `Optional[bool]` in this specific use case is perfectly fine. It's just that you should avoid it in general.

**Review by sebastian-philipp — APPROVED** [review]

File: src/pybind/ceph_argparse.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/ceph_argparse.py:173]: @epuertat after a second thought, i think we still cannot make `argdesc()` a classmethod, `CephChoices.__init__()` accepts `tp` which. and in future, i want to pass parameters like `range` and `goodchars` to the constructor of these `CephArgtype`s . currently, it's perfectly fine to make `argdesc()` a classmethod and decouple the `tp` stuff from a `CephChoices` instance, but that'd require us to pass `tp` (an Enum class) to `argdesc()`. that'd be inconsistent. so, i'd keep it this way. what do you think?

File: src/pybind/ceph_argparse.py [hunk 16]
```diff
@@ -147,9 +161,52 @@ def __str__(self):
...
+        elif tp is int:
+            return CephInt().argdesc(attrs)
+        elif tp is float:
+            return CephFloat().argdesc(attrs)
+        elif tp is bool:
+            return CephBool().argdesc(attrs)
+        elif isinstance(tp, CephArgtype):
+            return tp.argdesc(attrs)
+        elif isinstance(tp, type) and issubclass(tp, enum.Enum):
+            return CephChoices(enum=tp).argdesc(attrs)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/ceph_argparse.py:183]: thanks. incorporated with minor editing.

File: src/pybind/mgr/mgr_module.py (hunk 14 above)
Comment (tchaikov) [reply src/pybind/mgr/mgr_module.py:315]: @epuertat indeed! that's exactly what i've been thinking. but i'd like to leave it for another PR.

File: doc/_ext/ceph_commands.py [hunk 17]
```diff
@@ -156,31 +154,37 @@ def is_reasonably_simple(self):
...
         return line
 
 
 class Sig:
     @staticmethod
-    def _param_to_sig(p):
+    def _parse_arg_desc(desc):
         try:
-            return {kv.split('=')[0]: kv.split('=')[1] for kv in p.split(',') if kv}
+            return {kv.split('=')[0]: kv.split('=')[1] for kv in desc.split(',') if kv}
```
Comment (epuertat) [inline doc/_ext/ceph_commands.py:165]: If I'm not wrong, tuples are also valid for dict comprehensiones.

```suggestion
            return {kv.split('=') for kv in desc.split(',') if kv}
```

How strict is this command prefix syntax? Whitespaces are not allowed?: `k1=v1, k2 = v2 , ...`?

File: src/pybind/mgr/mgr_module.py (hunk 14 above)
Comment (epuertat) [reply src/pybind/mgr/mgr_module.py:315]: No problem... In fact this is less trivial than I initially though as we'd need access to the instantiated `MgrModule.module_name` property...

**Review by epuertat — COMMENTED** [review]
I'd suggest to add a brief comment to Mgr Module Doc: https://github.com/ceph/ceph/blob/master/doc/mgr/modules.rst

**Comment by tchaikov** [comment]
adding DNM, as there are a bunch of lint errors.

File: src/pybind/mgr/mgr_module.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/mgr_module.py:354]: @sebastian-philipp i thought about your suggestion again. and i think the beauty of it is that it makes it more explicit that we don't support Union actually. so i adapted it a little bit.

```py
     @staticmethod
     def _is_optional(tp):
         if get_origin(tp) is Union:
             if type(None) not in get_args(tp):
                 raise ValueError('Union type not supported in type annotations')
             return True
         else:
             return False
```

File: doc/_ext/ceph_commands.py (hunk 17 above)
Comment (tchaikov) [reply doc/_ext/ceph_commands.py:165]: @epuertat we do have a use case where there are space(s) after ",". see https://github.com/ceph/ceph/blob/72cc819b3aa47cdb684577be1f33021aa9fd2faf/src/mon/MonCommands.h#L515-L523

we could enforce the policy that only a single "," is allowed for the delimiter between parameter properties. but that's not in the scope of this PR. so i'd defer it to another change.

File: src/pybind/mgr/orchestrator/module.py (hunk 7 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator/module.py:1126]: @sebastian-philipp fixed anyway.

File: doc/_ext/ceph_commands.py [hunk 18]
```diff
@@ -156,31 +154,37 @@ def is_reasonably_simple(self):
...
 
 class Sig:
     @staticmethod
-    def _param_to_sig(p):
+    def _parse_arg_desc(desc):
         try:
-            return {kv.split('=')[0]: kv.split('=')[1] for kv in p.split(',') if kv}
-        except IndexError:
-            return p
+            return dict(kv.split('=') for kv in desc.split(',') if kv)
```
Comment (epuertat) [inline doc/_ext/ceph_commands.py:165]: Aha, so then if we want to clean trailing whitespaces, then we have to:
```suggestion
            return dict(map(str.strip, kv.split('=')) for kv in desc.split(',') if kv)
```

Comment (tchaikov) [reply doc/_ext/ceph_commands.py:165]: thanks. incorporated.

File: src/pybind/mgr/dashboard/plugins/debug.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/dashboard/plugins/debug.py:58]: no, otherwise I have 
```
File "ceph/src/pybind/mgr/.tox/mypy/lib/python3.9/site-packages/mypy/nodes.py", line 2660, in __getattribute__
    raise AssertionError(object.__getattribute__(self, 'msg'))
AssertionError: Var is lacking info
```
python3.6 has the same issue, and mypy 0.80 is able to finish the check. but it is not yet released at the time of writing.

**Comment by tchaikov** [comment]
@epuertat i documented the CLICommand in the last commit in this PR. could you please take a look?

**Comment by epuertat** [comment]
> @epuertat i documented the CLICommand in the last commit in this PR. see [ceph--38722.org.readthedocs.build/en/38722/mgr/modules](https://ceph--38722.org.readthedocs.build/en/38722/mgr/modules/) . could you please take a look?

Looks great, @kefu, and also opens the door to FTL travel discussions 🤣 Just a suggestion, in the sample snippet you put there, it'd be even more complete if you added the `return HandleCommandResult(retval, stdout, stderr)` statement, since from experience it is the first issue a devel hits when trying to implement their first CLI Command.

I'm sure @zdover23 will also have better suggestions on this! That document is the entry point for new ceph-mgr module developers, so I think it's crucial that we put as much love as we can to that! :D

**Comment by tchaikov** [comment]
@epuertat thanks. i made the FTL example more realistic, so it now returns a `HandleCommandResult`. =)

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-01-14_16:22:32-rados-wip-kefu-testing-2021-01-14-2241-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-15_15:58:16-rados-wip-kefu-testing-2021-01-15-1726-distro-basic-smithi/
